### ISSUE TITLE
Build fail for newer rack builds. 

### ISSUE THREAD

**Issue by Mubali** [VCVRack/Rack#301]
OS (if applicable): 10.10.5
Version (or "dev" if compiling from source):dev 

Cannot build from source. Gets following error during make:
In file included from src/core/MidiCCToCV.cpp:1:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/list:174:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:1617:20: error: 
object expression of non-scalar type 'bool [3]' cannot be used in a
pseudo-destructor expression

Not sure if this is an xcode issue or what, but I can't build any of the new ones with the sample rate change.

**Comment by bontric** [comment]
@AndrewBelt This is due to 
(MidiIO.hpp:48)
```c++
ignoresMap.erase(id);
```

[This](https://stackoverflow.com/questions/26307698/c-pseudo-destructor-on-array-type) seems to be somewhat related so I assume it's a problem to use a basic array type within an ordered map since `erase` probably tries to call a destructor . 

@Mubali could you try to replace 
(MidiIo.hpp:22)
```c
std::unordered_map<int, bool[3]> ignoresMap;
```
with
```c
std::unordered_map<int, std::array<bool, 3>> ignoresMap;
```
and see if this fixes your issue?

**Comment by Mubali** [comment]
Unfortunately, this does not fix my issues... Here are the new errors. I
sent the whole report just in case that might help.

In file included from src/core/MidiCCToCV.cpp:5:

*src/core/MidiIO.hpp:38:17: **error: **type 'mapped_type' (aka
'std::__1::array<bool,*

*      3>') does not provide a subscript operator*

                ignoresMap[id][0] = true;

*                ~~~~~~~~~~~~~~^~*

*src/core/MidiIO.hpp:39:17: **error: **type 'mapped_type' (aka
'std::__1::array<bool,*

*      3>') does not provide a subscript operator*

                ignoresMap[id][1] = true;

*                ~~~~~~~~~~~~~~^~*

*src/core/MidiIO.hpp:40:17: **error: **type 'mapped_type' (aka
'std::__1::array<bool,*

*      3>') does not provide a subscript operator*

                ignoresMap[id][2] = true;

*                ~~~~~~~~~~~~~~^~*

In file included from src/core/MidiCCToCV.cpp:1:

In file included from
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/list:174:

In file included from
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:602:

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:259:9:
**error: *

*      implicit instantiation of undefined template 'std::__1::array<bool,
3>'*

    _T2 second;

*        ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_map:540:16:
note: *

      in instantiation of template class 'std::__1::pair<const int,

      std::__1::array<bool, 3> >' requested here

    value_type __cc;

*               ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:59:16:
note: *

      in instantiation of template class 'std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >' requested here

    value_type __value_;

*               ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:228:22:
note: *

      in instantiation of template class

      'std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *>' requested here

    typedef typename __node::value_type                value_type;

*                     ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_map:654:19:
note: *

      in instantiation of template class


'std::__1::__hash_const_iterator<std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *> *>' requested here

    _HashIterator __i_;

*                  ^*

*src/core/MidiIO.hpp:48:20: note: *in instantiation of template class


'std::__1::__hash_map_const_iterator<std::__1::__hash_const_iterator<std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *> *> >' requested here

                ignoresMap.erase(id);

*                                 ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tuple:95:65:
note: *

      template is declared here

template <class _Tp, size_t _Size> struct _LIBCPP_TYPE_VIS_ONLY array;

*                                                                ^*

In file included from src/core/MidiCCToCV.cpp:1:

In file included from
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/list:174:

In file included from
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:602:

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:259:9:
**error: *

*      implicit instantiation of undefined template 'std::__1::array<bool,
3>'*

    _T2 second;

*        ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_map:541:21:
note: *

      in instantiation of template class 'std::__1::pair<int,

      std::__1::array<bool, 3> >' requested here

    __nc_value_type __nc;

*                    ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:59:16:
note: *

      in instantiation of template class 'std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >' requested here

    value_type __value_;

*               ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:228:22:
note: *

      in instantiation of template class

      'std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *>' requested here

    typedef typename __node::value_type                value_type;

*                     ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_map:654:19:
note: *

      in instantiation of template class


'std::__1::__hash_const_iterator<std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *> *>' requested here

    _HashIterator __i_;

*                  ^*

*src/core/MidiIO.hpp:48:20: note: *in instantiation of template class


'std::__1::__hash_map_const_iterator<std::__1::__hash_const_iterator<std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *> *> >' requested here

                ignoresMap.erase(id);

*                                 ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tuple:95:65:
note: *

      template is declared here

template <class _Tp, size_t _Size> struct _LIBCPP_TYPE_VIS_ONLY array;

*                                                                ^*

In file included from src/core/MidiCCToCV.cpp:5:

In file included from src/core/MidiIO.hpp:1:

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_map:657:43:
**error: *

*      'value_type' (aka 'int') is not a class, namespace, or enumeration*

    typedef const typename _HashIterator::value_type::value_type::first_...

*                                          ^*

*src/core/MidiIO.hpp:48:20: note: *in instantiation of template class


'std::__1::__hash_map_const_iterator<std::__1::__hash_const_iterator<std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *> *> >' requested here

                ignoresMap.erase(id);

*                                 ^*

In file included from src/core/MidiCCToCV.cpp:5:

In file included from src/core/MidiIO.hpp:1:

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_map:658:37:
**error: *

*      'value_type' (aka 'int') is not a class, namespace, or enumeration*

    typedef typename _HashIterator::value_type::value_type::second_typ...

*                                    ^*

In file included from src/core/MidiCCToCV.cpp:5:

In file included from src/core/MidiIO.hpp:1:

In file included from
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_map:350:

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:2183:14:
**error: *

*      no viable conversion from 'const_iterator' (aka*

*      '__hash_const_iterator<__node_pointer>') to 'iterator' (aka*

*      '__hash_iterator<__node_pointer>')*

    iterator __i = find(__k);

*             ^     ~~~~~~~~~*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/unordered_map:952:59:
note: *

      in instantiation of function template specialization

      'std::__1::__hash_table<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, std::__1::__unordered_map_hasher<int,

      std::__1::__hash_value_type<int, std::__1::array<bool, 3> >,

      std::__1::hash<int>, true>, std::__1::__unordered_map_equal<int,

      std::__1::__hash_value_type<int, std::__1::array<bool, 3> >,

      std::__1::equal_to<int>, true>,

      std::__1::allocator<std::__1::__hash_value_type<int,
std::__1::array<bool,

      3> > > >::__erase_unique<int>' requested here

    size_type erase(const key_type& __k) {return
__table_.__erase_unique(__k);}

*                                                          ^*

*src/core/MidiIO.hpp:48:14: note: *in instantiation of member function

      'std::__1::unordered_map<int, std::__1::array<bool, 3>,

      std::__1::hash<int>, std::__1::equal_to<int>,

      std::__1::allocator<std::__1::pair<const int, std::__1::array<bool,
3> > >

      >::erase' requested here

                ignoresMap.erase(id);

*                           ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:89:29:
note: *

      candidate constructor (the implicit copy constructor) not viable: no
known

      conversion from 'const_iterator' (aka

      '__hash_const_iterator<__node_pointer>') to 'const


std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *> *> &' for 1st argument

class _LIBCPP_TYPE_VIS_ONLY __hash_iterator

*                            ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:89:29:
note: *

      candidate constructor (the implicit move constructor) not viable: no
known

      conversion from 'const_iterator' (aka

      '__hash_const_iterator<__node_pointer>') to


'std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *> *> &&' for 1st argument

class _LIBCPP_TYPE_VIS_ONLY __hash_iterator

*                            ^*

*/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__hash_table:203:5:
note: *

      candidate constructor not viable: no known conversion from

      'const_iterator' (aka '__hash_const_iterator<__node_pointer>') to

      '__node_pointer' (aka

      'std::__1::__hash_node<std::__1::__hash_value_type<int,

      std::__1::array<bool, 3> >, void *> *') for 1st argument

    __hash_iterator(__node_pointer __node) _NOEXCEPT

*    ^*

8 errors generated.

make: *** [build/src/core/MidiCCToCV.cpp.o] Error 1

**Comment by bontric** [comment]
Okay, this looks like an issue with your build system. (Both options work fine for me and the std::_1::array stuff looks kind of weird). I assume that the wrong stdlib is linked. @AndrewBelt Do you have any idea?

As I'm not working with Mac/Xcode I have no experience with this. However it might be something similar to [this](https://stackoverflow.com/questions/4574246/can-i-use-c11-with-xcode). Maybe you can try that.

**Comment by AndrewBelt** [comment]
Don't use bool[3] or even std::array<bool, 3>. Think C-like with basic data structures like that. Why not move the functionality of `ignoresMap` into the individual Module subclasses?

**Comment by bontric** [comment]
> Don't use bool[3] or even std::array<bool, 3>. Think C-like with basic data structures.
 
Sorry but I really don't understand why those are problematic. You don't need to give me an explanation but it would be cool if you can point me somewhere. (I come more from a C/Go background and am a bit rusty when it comes to C++)

> Why not move the functionality of ignoresMap into the individual Module subclasses?

I had that at first but decided against it for some reason. Looking at it now, you are right. I can take care of that if you want

**Comment by AndrewBelt** [comment]
You guessed it correctly above. The type `bool[3]` cannot be assigned a destructor (or a constructor as well) which is required by std::unordered_map. `std::array<bool, 3>` can, but it's bad practice because you can just use

```cpp
struct IgnoreFlags {
	bool midiSysex;
	bool midiTime;
	bool midiSense;
}

std::unordered_map<int, IgnoreFlags> ignoresMap;
```

and as it's much more readable as a bonus.

**Comment by bontric** [comment]
Ahh, thanks! I completely misunderstood what you wrote :octopus: 

I will fix this and add it to #309 so you only need to merge once

**Comment by bontric** [comment]
@AndrewBelt I just discovered a major bug in the MidiIO implementation. I did some messy stuff with  the unordered_map's, but I'll add a fix for that as well.

**Comment by bontric** [comment]
@AndrewBelt #309 is done. Cleaned up the midiIO implementation. We still need the ignores map thoug, to make sure that we always have the right ignore types set even after modules are loaded.

**Comment by martin-lueders** [comment]
@bontric  Do you have any fix, I could do to compile on Mac, until the issue is repaired in Andrew's master branch?

**Comment by bontric** [comment]
@martin-lueders 
The fix is addressed in pull request #309, if you want to get it, do:
(Copy your rack folder and do this in the copy)

```bash
git pull https://github.com/bontric/Rack-MIDICC-Fork 2017-10-bontric-midiCC-fix
```
